File: Robot Control Course/Chapter5/Lesson1/Chapter5_Lesson1.java

public class PlanarArm {
    double l1 = 1.0, l2 = 0.7;
    double q1Min = -Math.PI/2.0, q1Max = Math.PI/2.0;
    double q2Min = -Math.PI,      q2Max = Math.PI;
    double xObs = 0.7, yObs = 0.3, rObs = 0.2;

    public double[] forwardKinematics(double[] q) {
        double q1 = q[0], q2 = q[1];
        double x = l1 * Math.cos(q1) + l2 * Math.cos(q1 + q2);
        double y = l1 * Math.sin(q1) + l2 * Math.sin(q1 + q2);
        return new double[]{x, y};
    }

    public boolean inJointLimits(double[] q) {
        double q1 = q[0], q2 = q[1];
        return (q1Min <= q1 && q1 <= q1Max) &&
               (q2Min <= q2 && q2 <= q2Max);
    }

    public double obstacleConstraint(double[] q) {
        double[] ee = forwardKinematics(q);
        double dx = ee[0] - xObs;
        double dy = ee[1] - yObs;
        return dx*dx + dy*dy - rObs*rObs; // >= 0 in free space
    }

    public double groundConstraint(double[] q) {
        double[] ee = forwardKinematics(q);
        return ee[1]; // y >= 0
    }

    public boolean admissible(double[] q) {
        if (!inJointLimits(q)) return false;
        if (obstacleConstraint(q) < 0.0) return false;
        if (groundConstraint(q) < 0.0) return false;
        return true;
    }

    public static void main(String[] args) {
        PlanarArm arm = new PlanarArm();
        double[] q = {0.0, 0.0};
        System.out.println("Admissible? " + arm.admissible(q));
    }
}

File: Robot_Control_Course/Chapter5/Lesson1/Chapter5_Lesson1.m

function ok = planar_constraint_check(q)
    % q: [q1; q2]
    l1 = 1.0; l2 = 0.7;
    q1_min = -pi/2; q1_max = pi/2;
    q2_min = -pi;   q2_max = pi;
    x_obs = 0.7; y_obs = 0.3; r_obs = 0.2;

    q1 = q(1); q2 = q(2);
    % Joint limits
    if q1 < q1_min || q1 > q1_max || q2 < q2_min || q2 > q2_max
        ok = false; return;
    end

    % FK
    x = l1*cos(q1) + l2*cos(q1 + q2);
    y = l1*sin(q1) + l2*sin(q1 + q2);

    % Obstacle constraint
    dx = x - x_obs; dy = y - y_obs;
    h = dx*dx + dy*dy - r_obs^2;  % should be >= 0
    if h < 0
        ok = false; return;
    end

    % Ground constraint y >= 0
    if y < 0
        ok = false; return;
    end

    ok = true;
end
